reconcile/utils/terrascript/cloudflare_client.py

```diff
@@ -146,6 +161,104 @@ def dumps(self) -> str:
         """Return the Terraform JSON representation of the resources"""
         return str(self._terrascript)
 
-    def _add_resources(self, tf_resources: Iterable[Union[Resource, Output]]) -> None:
+    def _add_resources(
+        self, tf_resources: Iterable[Union[Resource, Output, Data]]
+    ) -> None:
         for resource in tf_resources:
             self._terrascript.add(resource)
+
+
+class TerraformS3StateNamingStrategy(ABC):
+    @abstractmethod
+    def get_object_key(self, qr_integration: Integration) -> str:
+        pass
+
+
+class Default(TerraformS3StateNamingStrategy):
+    def get_object_key(self, qr_integration: Integration) -> str:
+        return qr_integration.key
+
+
+class AccountShardingStrategy(TerraformS3StateNamingStrategy):
+    def __init__(self, account: CloudflareAccount):
+        super().__init__()
+        self.account: CloudflareAccount = account
+
+    def get_object_key(self, qr_integration: Integration) -> str:
+        return f"{qr_integration.name}-{self.account.name}.tfstate"
+
+
+class TerrascriptCloudflareClientFactory:
+    @staticmethod
+    def _create_backend_config(
+        tf_state_s3: TerraformStateS3, key: str, secret_reader: SecretReader
+    ) -> TerraformS3BackendConfig:
+        aws_acct_creds = secret_reader.read_all(
+            {"path": tf_state_s3.automation_token_path}
+        )
```

Comment:
With the typed secret reader mentioned above (via `create_secret_reader()`) you can directly pass in a secret object and dont need to construct dictionaries any longer.

```python
aws_acct_creds = secret_reader.read_all_secret(tf_state_s3.automation_token)
```

This requires however, that the secret object is passed around instead of a path.